MoreLinq/PadLeft.cs

```diff
@@ -0,0 +1,162 @@
+        /// </code>
+        /// The <c>result</c> variable will contain <c>{ -1, -1, 123, 456, 789 }</c>.
+        /// </example>
+
+        public static IEnumerable<TSource> PadLeft<TSource>(this IEnumerable<TSource> source, int width, TSource padding)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (width < 0) throw new ArgumentException(null, nameof(width));
+            return PadLeftImpl(source, width, padding, null);
+        }
+
+        /// <summary>
+        /// Pads a sequence with a dynamic filler value in its beginning if it is narrower (shorter 
+        /// in length) than a given width.
+        /// An additional parameter specifies the function to calculate padding.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source"/>.</typeparam>
+        /// <param name="source">The sequence to pad.</param>
+        /// <param name="width">The width/length below which to pad.</param>
+        /// <param name="paddingSelector">Function to calculate padding.</param>
+        /// <returns>
+        /// Returns a sequence that is at least as wide/long as the width/length
+        /// specified by the <paramref name="width"/> parameter.
+        /// </returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+        /// <example>
+        /// <code>
+        /// int[] numbers = { 123, 456, 789 };
+        /// var result = numbers.PadLeft(6, i => -i);
+        /// </code>
+        /// The <c>result</c> variable will contain <c>{ 0, -1, -2, 123, 456, 789 }</c>.
+        /// </example>
+
+        public static IEnumerable<TSource> PadLeft<TSource>(this IEnumerable<TSource> source, int width, Func<int, TSource> paddingSelector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (paddingSelector == null) throw new ArgumentNullException(nameof(paddingSelector));
+            if (width < 0) throw new ArgumentException(null, nameof(width));
+            return PadLeftImpl(source, width, default(TSource), paddingSelector);
+        }
+
+        private static IEnumerable<T> PadLeftImpl<T>(IEnumerable<T> source,
+            int width, T padding, Func<int, T> paddingSelector)
+        {
+            if (source is ICollection<T> col)
+            {
+                if (col.Count < width) 
+                    return Enumerable.Range(0, width - col.Count)
+                                     .Select((x, i) => paddingSelector != null ? paddingSelector(i) : padding)
+                                     .Concat(col);
+                
+                return col;
+            }
+
+            return _(); IEnumerable<T> _()
+            {
+                using (var e = source.GetEnumerator())
+                {
+                    var list = new List<T>(width);
```

Comment:
Why not use an array instead of a `List<T>` since it is always going to be a fixed size? It will also save an extra allocation.